### PR TITLE
Accelerated checkins

### PR DESCRIPTION
**Pull request by jedi22**
Added a setting allowing the server to request that osquery checkin every 5 seconds for a length of time that it specifies.

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

File: osquery/dispatcher/distributed.cpp [hunk 1]
```diff
@@ -30,7 +32,16 @@ void DistributedRunner::start() {
     if (dist.getPendingQueryCount() > 0) {
       dist.runQueries();
     }
-    pauseMilli(FLAGS_distributed_interval * 1000);
+    std::string str_acu = "0";
+    Status stat = getDatabaseValue(kPersistentSettings,
+        "accelerate_checkins_until",
+        str_acu);
+    unsigned long accelerate_checkins_until = std::stol(str_acu);
```
Comment (PoppySeedPlehzr) [inline osquery/dispatcher/distributed.cpp]: Try to make use of `auto` where applicable, as opposed to `unsigned long`

File: osquery/distributed/distributed.cpp [hunk 2]
```diff
@@ -182,6 +183,15 @@ Status Distributed::acceptWork(const std::string& work) {
       }
       setDatabaseValue(kQueries, kDistributedQueryPrefix + node.first, query);
     }
+    if (tree.count("accelerate_checkins_for") > 0) {
+      auto new_time =
+          tree.get<std::string>("accelerate_checkins_for", "");
+      unsigned long duration = std::stol(new_time);
```
Comment (PoppySeedPlehzr) [inline osquery/distributed/distributed.cpp]: `auto`

**Comment by jedi22** [comment]
ok to test

**Comment by PoppySeedPlehzr** [comment]
ok to test

**Comment by osqueryer** [comment]
:no_good: The commit 972abb02a646ee05782e554d7adae5ef7b746f8a (Job results: [206](https://jenkins.osquery.io/job/osqueryPullRequestAudit/206/)) failed the code audit.

File: osquery/distributed/distributed.cpp [hunk 3]
```diff
@@ -182,6 +183,15 @@ Status Distributed::acceptWork(const std::string& work) {
       }
       setDatabaseValue(kQueries, kDistributedQueryPrefix + node.first, query);
     }
+    if (tree.count("accelerate_checkins_for") > 0) {
+      auto new_time = tree.get<std::string>("accelerate_checkins_for", "");
+      auto duration = std::stol(new_time);
+      LOG(INFO) << "Accelerating Checkins for " << duration << " seconds.\n";
```
Comment (theopolis) [inline osquery/distributed/distributed.cpp]:
```cpp
LOG(INFO) << "Accelerating distributed query checkins for " << duration << " seconds";
```

File: osquery/dispatcher/distributed.cpp [hunk 4]
```diff
@@ -30,7 +32,15 @@ void DistributedRunner::start() {
     if (dist.getPendingQueryCount() > 0) {
       dist.runQueries();
     }
-    pauseMilli(FLAGS_distributed_interval * 1000);
+    std::string str_acu = "0";
+    Status stat = getDatabaseValue(
+        kPersistentSettings, "accelerate_checkins_until", str_acu);
```
Comment (theopolis) [inline osquery/dispatcher/distributed.cpp]: `distributed_accelerate_checkins_expire`.

File: osquery/dispatcher/distributed.cpp [hunk 5]
```diff
@@ -30,7 +32,15 @@ void DistributedRunner::start() {
...
+    auto accelerate_checkins_until = std::stol(str_acu);
```
Comment (theopolis) [inline osquery/dispatcher/distributed.cpp]: Take a look at our `safeStol*` methods in `./core/conversions.h`.

File: osquery/dispatcher/distributed.cpp [hunk 6]
```diff
@@ -30,7 +32,15 @@ void DistributedRunner::start() {
...
+    if (!stat.ok() || getUnixTime() > accelerate_checkins_until) {
+      pauseMilli(FLAGS_distributed_interval * 1000);
+    } else {
+      pauseMilli(5000);
```
Comment (theopolis) [inline osquery/dispatcher/distributed.cpp]: Maybe define:
```cpp
const size_t kDistributedAccelerationInterval = 5;
```

File: osquery/distributed/distributed.cpp [hunk 7]
```diff
@@ -182,6 +183,15 @@ Status Distributed::acceptWork(const std::string& work) {
       }
       setDatabaseValue(kQueries, kDistributedQueryPrefix + node.first, query);
     }
+    if (tree.count("accelerate_checkins_for") > 0) {
+      auto new_time = tree.get<std::string>("accelerate_checkins_for", "");
+      auto duration = std::stol(new_time);
```
Comment (theopolis) [inline osquery/distributed/distributed.cpp]: Same with `safe*` methods.

File: tools/tests/test_http_server.py [hunk 8]
```diff
@@ -37,6 +37,13 @@
     }
 }
 
+EXAMPLE_DISTRIBUTED_ACCELERATE = {
+    "queries": {
+        "info": "select * from osquery_info",
+    },
+    "accelerate_checkins_for" : "60"
```
Comment (theopolis) [inline tools/tests/test_http_server.py]: Can we pick another key? `accelerate` maybe?

**Comment by facebook-github-bot-4** [comment]
@jedi22 updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2454/files/34ad1a3cfe10fb5cc3ef0526cb826164da43b7bd..d9803290455b945c3f5082cb5eb9bb11b39330a4)

**Comment by osqueryer** [comment]
:-1: The commit 0a5dd29a18050a1ceafc37fdcfc539eb7a7e134a (Job results: [3225](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3225/)) failed one or more tests (**OS X/Linux**).

**Comment by facebook-github-bot-1** [comment]
@jedi22 updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2454/files/d9803290455b945c3f5082cb5eb9bb11b39330a4..34ad1a3cfe10fb5cc3ef0526cb826164da43b7bd)

**Comment by facebook-github-bot-7** [comment]
@jedi22 updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2454/files/34ad1a3cfe10fb5cc3ef0526cb826164da43b7bd..972abb02a646ee05782e554d7adae5ef7b746f8a)

**Comment by facebook-github-bot-9** [comment]
@jedi22 updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2454/files/972abb02a646ee05782e554d7adae5ef7b746f8a..d9803290455b945c3f5082cb5eb9bb11b39330a4)

**Comment by facebook-github-bot-2** [comment]
@jedi22 updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2454/files/d9803290455b945c3f5082cb5eb9bb11b39330a4..0a5dd29a18050a1ceafc37fdcfc539eb7a7e134a)

**Comment by facebook-github-bot-1** [comment]
@jedi22 updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2454/files/0a5dd29a18050a1ceafc37fdcfc539eb7a7e134a..10f230a14c9d893a80b26a8f6f65aef63f139a5d)

**Comment by facebook-github-bot-0** [comment]
@jedi22 updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2454/files/10f230a14c9d893a80b26a8f6f65aef63f139a5d..00e71d26355ff3b1441bfb454db04c809cff918e)

**Comment by facebook-github-bot-9** [comment]
@jedi22 updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2454/files/00e71d26355ff3b1441bfb454db04c809cff918e..dc8a6b90eb1337a5b67c3f6948425e9cdf83d90f)

**Pull request merged**